co-simulation tool/docker/ns3-docker/user-guide-docker.md

```diff
@@ -0,0 +1,222 @@
+
+```
+docker run --rm -it --gpus all -v /tmp/.X11-unix:/tmp/.X11-unix -e DISPLAY=$DISPLAY -e QT_X11_NO_MITSHM=1 --user=carla_sumo_mosaic ns-3-integration
+```
+
+### Modify files in Docker Container
+
+#### Step 1: Modify `ns3_installer.sh`
+
+In Docker container, type:
+
+```
+cd ns-3-integration/bin/fed/ns3
+```
+
+Then modify `ns3_installer.sh` by:
+
+```
+vim ns3_installer.sh
+```
+
+(If you see a message `vim: command not found`, please use `sudo apt-get install vim` to install vim as text editor)
+
+In line 365, change
+
+```
+cp -ar ns-3.28/build/ns3 /usr/include/
+```
+
+to 
+
+```
+sudo cp -ar ns-3.28/build/ns3 /usr/include/
+```
+
+#### Step 2: Build NS-3
+
+In directory `ns-3-integration/bin/fed/ns3`, type:
+
+```
+./ns3_installer.sh
+```
+
+Once all 1947 files are compiled and a successful message shows up, it means NS-3 is successful installed.
+
+#### Step 3: Add NS-3 as environment variable
+
+```
+export NS3_HOME='/ns-3-integration/bin/fed/ns3'
+```
+
+Use `printenv` to check if NS-3 is successful added to environment variable.
+
+#### Step 4: Modify `run.sh`
+
+In directory `ns-3-integration/bin/fed/ns3`, type:
+
+```
+vim run.sh
+```
+
+Change line 11 to:
+
+```
+LD_LIBRARY_PATH=/ns-3-integration/fed/bin/ns3/ns-allinone-3.28/ns-3.28/build /ns-3-integration/fed/bin/ns3/ns-allinone-3.28/ns-3.28/build/scratch/mosaic_starter --port=$port --cmdPort=$cmdport --configFile=scratch/ns3_federate_config.xml
+```
+
+Now, you can test if NS-3 works.
+
+In root directory, type:
+
+```
+cd ns-3-integration
+```
+
+Use example `Tiergarten` to test:
+
+```
+./mosaic.sh -s Tiergarten
+```
+
+#### Step 5: Modify traci library (For Co-Simulation)
+
+In root directory, type:
+
+```
+cd /usr/share/sumo/tools/traci
+```
+
+Then in the directory `/usr/share/sumo/tools/traci`, modify the following three files:
+
+1. `constants.py`
+
+   ```
+   sudo nano constants.py
+   ```
+
+   Add these to `constants.py`:
+
+   ```python
+   # command: get V2X message
+   CMD_GET_V2X = 0x0d
+   
+   # command: set V2X message
+   CMD_SET_V2X = 0x2f
+   ```
+
+2. `main.py`
+
+   ```
+   sudo nano main.py
+   ```
+
+   Add these methods to `main.py`:
+
+   ```python
+   def getV2xMessage():
+       if "" not in _connections:
+           raise FatalTraCIError("Not connected.")
+       return _connections[""].getV2xMessage()
+   
+   def setV2xMessage(message):
+       if "" not in _connections:
+           raise FatalTraCIError("Not connected.")
+       return _connections[""].setV2xMessage(message)
+   ```
+
+3. `connection.py`
+
+   ```
+   sudo nano connection.py
+   ```
+
+   Add the following commands:
+
+   ```python
+   def getV2xMessage(self):
+       command = tc.CMD_GET_V2X
+       result = self._sendCmd(command,None,None)
+       result.readLength()
+       response = result.read("!B")[0]
+   
+       if response != command:
+           raise FatalTraCIError("Received answer %s for command %s." % (response, command))
+   	return result.readStringList()
```

Comment:
Probably adding one more tab here. Return method in the viewing mode is inside of condition method. Try to put that outside of condition method

Reply:
The statement has been indented as suggested.